### PR TITLE
refactor: Improve assert short code tooling

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Based on #13963.

The assert short code policy checker has been refactored to report all errors at once in an easy to read format, instead of only reporting the first error then existing. It also prints the source locations in a more vscode friendly format and factors out the policy for what functions to process into a map.

Also the documentation on assert is clarified/corrected,

This is an initial non-breaking step toward a better assert tagging system, and offers value on its own, but also eases possible future work.
Future work could build on this to extend the set of functions that can be processes and/or generalize it so it makes sense to run it over individual packages or release groups instead of being hard coded to this repo and specific function names repo wide.

To keep decisions about what to do in the future, as well as breaking changes out of this PR, the changes are rather minimal, which should make it easier to review and merge.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts [hunk 1]
```diff
@@ -133,29 +138,59 @@ export const handler: Handler = {
...
+			);
 		}
 
 		// go through all the newly collected asserts and add short codes
 		for (const s of newAssetFiles) {
 			// another policy may have changed the file, so reload it
 			s.refreshFromFileSystemSync();
 			for (const msg of getAssertMessageParams(s)) {
-				// here we only want to looks at those messages that are not numbers,
+				// here we only want to looks at those messages that are strings,
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts:191]:
```suggestion
				// here we only want to look at those messages that are strings,
```

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts [hunk 2]
```diff
@@ -209,7 +244,7 @@ function writeShortCodeMappingFile() {
 
 // Auto-generated by policy-check in @fluidframework/build-tools.
 
-export const shortCodeMap = ${JSON.stringify(mapContents, null, 4)};
+export const shortCodeMap = ${JSON.stringify(mapContents, null, "\t")};
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts:247]: nit: ideally whitespace would be configurable, but as you noted elsewhere, this is hardly the only thing that would better live in a config somewhere.

File: common/lib/common-utils/src/assert.ts [hunk 3]
```diff
@@ -4,11 +4,12 @@
  */
 
 /**
- * A browser friendly version of the node assert library. Use this instead of the 'assert' package, which has a big
- * impact on bundle sizes.
+ * A browser friendly assert library.
+ * Use this instead of the 'assert' package, which has a big impact on bundle sizes.
  * @param condition - The condition that should be true, if the condition is false an error will be thrown.
+ * Only use this API when `false` indicates an logic error in the problem and thus a bug that should be fixed.
```
Comment (tylerbutler) [inline common/lib/common-utils/src/assert.ts:10]:
```suggestion
 * Only use this API when `false` indicates a logic error in the problem and thus a bug that should be fixed.
```

**Review by tylerbutler — APPROVED** [review]
Very minor nits. Thanks for improving this!

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts [hunk 4]
```diff
@@ -14,42 +22,35 @@ const codeToMsgMap = new Map<string, string>();
...
- * Given a source file this function will look for all assert functions contained in it, and return the second parameter from
+ * Map from assertion function name to the index of its message argument.
+ *
+ * TODO:
+ * This should be moved into a configuration file.
+ */
+const assertionFunctions: ReadonlyMap<string, number> = new Map([["assert", 1]]);
+
+/**
+ * Given a source file this function will look for all assert functions contained in it, and return the message parameters.
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/assertShortCode.ts:38]: Nit
```suggestion
 * Given a source file, this function will look for all assert functions contained in it and return the message parameters.
```

**Pull request merged**